### ISSUE TITLE
LOS

### ISSUE THREAD

**Issue by c4c1n6kr3m1** [OpenKore/openkore#1687]
###### Affected configuration(s)/ file(s):
```Include relevant code snippets and source code references```
###### Impact: 
stuck

###### Expected Behavior:
not stuck

###### Steps to Reproduce:
use ranged attack

###### Additional Details:

bad LOS compare to client/server?
using warlock, ranger or any ranged attack, bot get stuck sometimes
**checkLineWalkable return 1**
but actually, because i used xkore , i can see there is still part of an obstacle between bot and targeted monster
so, we really have problem with LOS ?

**Comment by Nipodemos** [comment]
yes we do have a problem, but i have no ideia which one it is
i saw happening a few times, if there is an obstacle between monster and you, he tries to attack the monster infinitly, and the only way do make him get out of this loop is moving manually

**Comment by c4c1n6kr3m1** [comment]
you can check  the code , 
Misc.pm , when chosing target is LOS target or not
```			if (!checkLineWalkable($myPos, $pos)) {
				push(@noLOSMonsters, $_);
				next;
			}
```
attack.pm, when bot with range attack choose attacking spot potition with LOS with target 
```		for my $spot (@stand) {
			# Is this spot acceptable?
			# 1. It must have LOS to the target ($realMonsterPos).
			# 2. It must be within $config{followDistanceMax} of
			#    $masterPos, if we have a master.
			if (
			    (($config{attackCanSnipe} && checkLineSnipable($spot, $realMonsterPos))
				|| checkLineWalkable($spot, $realMonsterPos))
				&& $field->isWalkable($spot->{x}, $spot->{y})
				&& ($realMyPos->{x} != $spot->{x} && $realMyPos->{y} != $spot->{y})
				&& (!$master || round(distance($spot, $masterPos)) <= $config{followDistanceMax})
			) {
				my $dist = distance($realMyPos, $spot);
				if (!defined($best_dist) || $dist < $best_dist) {
					$best_dist = $dist;
					$best_spot = $spot;
				}
			}
		}
```
range attack with $config{attackCheckLOS} will check every walkable spot has LOS to target than choose the best ‎position
so if the LOS checking code (which is sub checkLineWalkable , Misc.pm) is perfect , this stuck problem should never happen

short story,
my warlock got stuck , i was using xkore ,  i can see a part of an obstacle (may be only a little tip ) between bot and targeted, 
, i try to attack using client, client reject it because it is not LOS spot
than i calculate using checkLineWalkable , between monster and the choosen spot( my bot already stand at this choosen spot)
than realize that ****checkLineWalkable** return 1** which mean, bot think this spot has LOS to target

this problem is different with https://github.com/OpenKore/openkore/issues/219

**Comment by Fadreus** [comment]
I think all ranged skill, including heal & buff also affected; not just attack ranged.
When ever I had my slave follow me (non bot) although the distance is less than 8 but LOS is not clear, it will stuck there attempting to cast skill until I pick her up (closing our distance or get clear LOS).

In some cases, even when the distance is between 0~0.9 it still won't cast skills unless target moved.

**Comment by c4c1n6kr3m1** [comment]
actually partySkill_ works different compare than attack skill
partySkill does not check LOS , so your slave always think master is in LOS, which actually not
```
In some cases, even when the distance is between 0~0.9 or at a plain open field, it still won't cast skills unless target moved.
```
i know we have this problem , but i never dig it
may be someone have clue about this problem

**Comment by c4c1n6kr3m1** [comment]
is any one really understand how LOS work in openkore , (sub checkLineWalkable - the math calculation ) , ?
fast reading... looks complicated for me

while i am thingking that RO server calculation is not as complicated as bot calculation
only using basic math 
```
y = mx + c
```
we have 2 pos , (x1,y1) and (x2,y2)
find the m
find the c
apply the formula to every single spot between x1 to x2
if any (x,y) between,... is unwalkable, than it is not Straight Line

it's night here ....
bug label anyone ?

**Comment by Nipodemos** [comment]
@Henrybk knows about this
in fact, i think his PR of pathfinding is dealing with this

**Comment by alisonrag** [comment]
fixed at #2871
